### PR TITLE
planner: support hash v2 variable to control the join order

### LINKED ISSUE #58325: Choose left side as build in semi and anti semi join in optimizer
## Enhancement

Currently, hash join can only choose right side as build table in semi and anti-semi join. For example, for the sql `select * from t1 where exists (select 1 from t2 where t1.col0 = t2.col0) order by t1.col0, t1.col1;`, if t2 has 1000w rows, t1 has 1000 rows. It's obvious that choose t1 as build table is better, but we don't support to build table for left side in previous hash join implementation. After [#57658](https://github.com/pingcap/tidb/pull/57658) and [#57971](https://github.com/pingcap/tidb/pull/57971) merge, hash join v2 could support left side as build table.  So we need to make optimizer enable to choose left side as build table when necessary.

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #58325

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/physical_plans.go [hunk 1]
```diff
@@ -1478,26 +1478,7 @@ type PhysicalHashJoin struct {
...
-	switch p.JoinType {
-	case logicalop.LeftOuterJoin, logicalop.RightOuterJoin, logicalop.InnerJoin:
-		return true
-	default:
-		return false
-	}
+	return isGAForHashJoinV2(p.JoinType, p.LeftJoinKeys, p.IsNullEQ, p.LeftNAJoinKeys)
 }
 
 // CanUseHashJoinV2 returns true if current join is supported by hash join v2
```
Comment (windtalker) [inline pkg/planner/core/physical_plans.go:1484]: Looks like this function has the same logical of `func canUseHashJoinV2(p *logicalop.LogicalJoin) bool`, I think we should share the common code?

Comment (hawkingrei) [reply pkg/planner/core/physical_plans.go:1483]: OK, I have merged those code.

**Review by windtalker — APPROVED** [review]
lgtm

**Review by AilinKid — APPROVED** [review]

**Pull request merged**